components/script/script_thread.rs

```diff
@@ -2073,12 +2075,17 @@ impl ScriptThread {
         window.evaluate_media_queries_and_report_changes();
     }
 
-    /// Initiate a non-blocking fetch for a specified resource. Stores the InProgressLoad
+    /// Sends a message to constellation to check for cross-origin redirects. Stores the InProgressLoad
     /// argument until a notification is received that the fetch is complete.
-    fn start_page_load(&self, incomplete: InProgressLoad, mut load_data: LoadData) {
+    fn pre_page_load(&self, incomplete: InProgressLoad, load_data: LoadData) {
         let id = incomplete.pipeline_id.clone();
+        self.constellation_chan.send(ConstellationMsg::ManualRedirect(id, load_data)).unwrap();
+        self.incomplete_loads.borrow_mut().push(incomplete);
+    }
 
-        let context = Arc::new(Mutex::new(ParserContext::new(id, load_data.url.clone())));
+    /// Initiate a non-blocking fetch for a specified resource.
+    fn start_page_load(&self, id: PipelineId, mut req_init: RequestInit) {
+        let context = Arc::new(Mutex::new(ParserContext::new(id, req_init.url.clone())));
```

Comment:
This is where we should be able to invoke the FetchResponseListener methods using the data provided by the constellation.